src/NHibernate/Dialect/Dialect.cs

```diff
@@ -101,7 +101,7 @@ protected Dialect()
 			RegisterFunction("cast", new CastFunction());
 			RegisterFunction("transparentcast", new TransparentCastFunction());
 			RegisterFunction("extract", new AnsiExtractFunction());
-			RegisterFunction("concat", new VarArgsSQLFunction(NHibernateUtil.String, "(", " || ", ")"));
+			RegisterFunction("concat", new VarArgsSQLFunction("(", " || ", ")"));
```

Comment:
I think this is an invalid change. It is not the first parameter which matters, but all parameters. As pointed in my first comment, all parameters must be ANSI strings for the return type to be an ANSI string. By not specifying the return type, we let NHibernate infer the returned type from the first parameter, which is then incorrect if the first is an ANSI string but not all others.

It is also worst to incorrectly type something as an ANSI string than incorrectly type something as an Unicode string: the former may cause garbled text and performance issues, while the later may only cause performance issues. (Well, since with SQL Server 2019, varchar can be [UTF-8 encoded](https://docs.microsoft.com/en-us/sql/relational-databases/collations/collation-and-unicode-support#utf8), the former in such case could not have the garbled text issue. But that remains a corner case for now I think.)

So I think this change is a regression.
We would need instead a specialized concat function which check the type of all its arguments, and would yield an ANSI string only if all of them are ANSI strings, otherwise it would yield an Unicode string.

Reply:
In case of simple comparison `p.AnsiColumn = @param` if `@param` type is not explicitly specified it will be treated as ANSI. Even though `@param` might contain some UTF-8 symbols...  

To me this case is not much of a difference. It's just an assumption that's used for type detection for not specified explicitly parameter types. So I don't think any garbled text is possible here (only wrong comparisons results, the same as with simple comparison)

But OK let's be on a safe side...